File: MobileOpsConnect/find_links.js
const fs = require('fs');
const path = require('path');

function getControllersAndActions() {
    const validRoutes = {};
    const controllerDir = 'Controllers';
    if (!fs.existsSync(controllerDir)) return validRoutes;
    
    fs.readdirSync(controllerDir).forEach(filename => {
        if (filename.endsWith('Controller.cs')) {
            const controllerName = filename.replace('Controller.cs', '');
            const filepath = path.join(controllerDir, filename);
            const content = fs.readFileSync(filepath, 'utf8');
            
            // Match: public [async Task<]IActionResult[>] MethodName[<...>] (
            const pattern = /public\s+(?:async\s+Task<)?(?:IActionResult|ActionResult)(?:>)?\s+(\w+)/g;
            const actions = new Set();
            let match;
            while ((match = pattern.exec(content)) !== null) {
                actions.add(match[1]);
            }
            validRoutes[controllerName] = Array.from(actions);
        }
    });
    return validRoutes;
}

function scanViewsForLinks() {
    const links = [];
    const viewDir = 'Views';
    if (!fs.existsSync(viewDir)) return links;
    
    function walk(dir) {
        fs.readdirSync(dir).forEach(file => {
            const fullPath = path.join(dir, file);
            if (fs.statSync(fullPath).isDirectory()) {
                walk(fullPath);
            } else if (file.endsWith('.cshtml')) {
                const content = fs.readFileSync(fullPath, 'utf8');
                
                // Extremely crude tag matching
                // just find pairs of asp-controller="..." asp-action="..."
                const tagPattern = /<[^>]+>/g;
                let match;
                while ((match = tagPattern.exec(content)) !== null) {
                    const tag = match[0];
                    const ctrlMatch = tag.match(/asp-controller="([^"]+)"/);
                    const actMatch = tag.match(/asp-action="([^"]+)"/);
                    
                    if (ctrlMatch && actMatch) {
                        links.push({
                            type: 'asp-helper',
                            controller: ctrlMatch[1],
                            action: actMatch[1],
                            file: fullPath
                        });
                    } else if (actMatch && !ctrlMatch) {
                        // Inherit controller from folder
                        const folder = path.basename(path.dirname(fullPath));
                        if (folder !== 'Shared') {
                            links.push({
                                type: 'asp-helper-implicit',
                                controller: folder,
                                action: actMatch[1],
                                file: fullPath
                            });
                        }
                    }
                }
                
                // Find hrefs
                const hrefPattern = /href="(\/?[^"]+)"/g;
                while ((match = hrefPattern.exec(content)) !== null) {
                    const url = match[1];
                    if (!url.startsWith('#') && !url.startsWith('http') && !url.startsWith('mailto:')) {
                        links.push({
                            type: 'href',
                            url: url,
                            file: fullPath
                        });
                    }
                }
            }
        });
    }
    walk(viewDir);
    return links;
}

function scanControllersForRedirects() {
    const redirects = [];
    const controllerDir = 'Controllers';
    if (!fs.existsSync(controllerDir)) return redirects;
    
    fs.readdirSync(controllerDir).forEach(filename => {
        if (filename.endsWith('Controller.cs')) {
            const filepath = path.join(controllerDir, filename);
            const content = fs.readFileSync(filepath, 'utf8');
            
            const p1 = /RedirectToAction\(\s*"([^"]+)"\s*,\s*"([^"]+)"\s*\)/g;
            let match;
            while ((match = p1.exec(content)) !== null) {
                redirects.push({
                    type: 'redirect',
                    controller: match[2],
                    action: match[1],
                    file: filepath
                });
            }
            
            const currentCtrl = filename.replace('Controller.cs', '');
            const p2 = /RedirectToAction\(\s*"([^"]+)"\s*\)/g;
            while ((match = p2.exec(content)) !== null) {
                redirects.push({
                    type: 'redirect',
                    controller: currentCtrl,
                    action: match[1],
                    file: filepath
                });
            }
        }
    });
    return redirects;
}

function analyze() {
    const routes = getControllersAndActions();
    const links = scanViewsForLinks();
    const redirects = scanControllersForRedirects();
    
    const errors = new Set();
    
    links.forEach(link => {
        if (link.type.startsWith('asp-helper')) {
            const c = link.controller;
            const a = link.action;
            if (!routes[c]) {
                errors.add(`404: Controller '${c}' not found. Referenced in ${link.file} (Action: ${a})`);
            } else if (!routes[c].includes(a)) {
                // Ignore generated Index vs empty, etc.
                errors.add(`404: Action '${a}' not found in Controller '${c}'. Referenced in ${link.file}`);
            }
        } else if (link.type === 'href') {
            let url = link.url.split('?')[0]; // strip query
            if (url.startsWith('/')) url = url.substring(1);
            const parts = url.split('/').filter(p => p !== '');
            if (parts.length >= 2) {
                const c = parts[0];
                const a = parts[1];
                // basic casing check
                // This might be flaky, so we just check if controller exists and action exists case-insensitively
                const matchedCtrlKey = Object.keys(routes).find(key => key.toLowerCase() === c.toLowerCase());
                if (matchedCtrlKey) {
                    const matchedAct = routes[matchedCtrlKey].find(act => act.toLowerCase() === a.toLowerCase());
                    if (!matchedAct) {
                        errors.add(`Possible 404: Href '/${c}/${a}' -> Action '${a}' not found in Controller '${matchedCtrlKey}'. Referenced in ${link.file}`);
                    }
                }
            }
        }
    });
    
    redirects.forEach(link => {
        const c = link.controller;
        const a = link.action;
        if (!routes[c]) {
            errors.add(`404: Redirect Controller '${c}' not found. Referenced in ${link.file} (Action: ${a})`);
        } else if (!routes[c].includes(a)) {
            errors.add(`404: Redirect Action '${a}' not found in Controller '${c}'. Referenced in ${link.file}`);
        }
    });

    errors.forEach(e => console.log(e));
}

analyze();

File: MobileOpsConnect/find_links.py
import os
import re

def get_controllers_and_actions():
    valid_routes = {}
    controller_dir = 'Controllers'
    if not os.path.exists(controller_dir):
        return valid_routes
        
    for filename in os.listdir(controller_dir):
        if filename.endswith('Controller.cs'):
            controller_name = filename.replace('Controller.cs', '')
            filepath = os.path.join(controller_dir, filename)
            with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
                # Find methods returning [Task<]IActionResult[>]
                # Match: public [async Task<]IActionResult[>] MethodName[<...>] (
                pattern = r'public\s+(?:async\s+Task<)?(?:IActionResult|ActionResult)(?:>)?\s+(\w+)'
                actions = set(re.findall(pattern, content))
                valid_routes[controller_name] = actions
    return valid_routes

def scan_views_for_links():
    links = []
    view_dir = 'Views'
    if not os.path.exists(view_dir):
        return links
        
    for root, _, files in os.walk(view_dir):
        for filename in files:
            if filename.endswith('.cshtml'):
                filepath = os.path.join(root, filename)
                with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
                    
                    # Match asp-controller and asp-action
                    # Sometimes they are on the same line, sometimes not
                    # Easiest way is to find tag by tag
                    tag_pattern = r'<a[^>]+>'
                    form_pattern = r'<form[^>]+>'
                    tags = re.findall(tag_pattern, content) + re.findall(form_pattern, content)
                    for tag in tags:
                        controller_match = re.search(r'asp-controller="([^"]+)"', tag)
                        action_match = re.search(r'asp-action="([^"]+)"', tag)
                        
                        if controller_match and action_match:
                            links.append({
                                'type': 'asp-helper',
                                'controller': controller_match.group(1),
                                'action': action_match.group(1),
                                'file': filepath
                            })
                        elif action_match and not controller_match:
                            # Assume same controller as folder name
                            folder = os.path.basename(os.path.dirname(root)) # Wait, root is usually Views/ControllerName
                            if os.path.basename(root) != 'Shared':
                                links.append({
                                    'type': 'asp-helper-implicit',
                                    'controller': os.path.basename(root),
                                    'action': action_match.group(1),
                                    'file': filepath
                                })
                                
                    href_pattern = r'href="(/?[^"]+)"'
                    actions_pattern = r'action="(/?[^"]+)"'
                    for href in re.findall(href_pattern, content) + re.findall(actions_pattern, content):
                        if href.startswith('#') or href.startswith('http') or href.startswith('mailto:'):
                            continue
                        links.append({
                            'type': 'href',
                            'url': href,
                            'file': filepath
                        })
    return links
    
def scan_controllers_for_redirects():
    redirects = []
    controller_dir = 'Controllers'
    if not os.path.exists(controller_dir):
        return redirects
        
    for filename in os.listdir(controller_dir):
        if filename.endswith('Controller.cs'):
            filepath = os.path.join(controller_dir, filename)
            with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
                # RedirectToAction("Action", "Controller")
                pattern1 = r'RedirectToAction\(\s*"([^"]+)"\s*,\s*"([^"]+)"\s*\)'
                # RedirectToAction("Action")
                pattern2 = r'RedirectToAction\(\s*"([^"]+)"\s*\)'
                for act, ctrl in re.findall(pattern1, content):
                    redirects.append({
                        'type': 'redirect',
                        'controller': ctrl,
                        'action': act,
                        'file': filepath
                    })
                current_ctrl = filename.replace('Controller.cs', '')
                for act in re.findall(pattern2, content):
                    redirects.append({
                        'type': 'redirect',
                        'controller': current_ctrl,
                        'action': act,
                        'file': filepath
                    })
    return redirects

def analyze():
    # Ignore identity/default Razor pages if any, but we primarily check MVC
    routes = get_controllers_and_actions()
    links = scan_views_for_links()
    redirects = scan_controllers_for_redirects()
    
    errors = []
    
    # Check ASP links
    for link in links:
        if link['type'] in ('asp-helper', 'asp-helper-implicit'):
            c = link['controller']
            a = link['action']
            if c not in routes:
                errors.append(f"404: Controller '{c}' not found. Referenced in {link['file']} (Action: {a})")
            elif a not in routes[c]:
                # Action methods can be [HttpGet], etc. or overloaded.
                # Also fallback to check if there is an attribute route?
                # This simple script might miss attribute routing, but MVC is usually convention-based
                errors.append(f"404: Action '{a}' not found in Controller '{c}'. Referenced in {link['file']}")
                
        elif link['type'] == 'href':
            href = link['url']
            # simple /Controller/Action
            parts = [p for p in href.split('/') if p and '?' not in p]
            if len(parts) >= 2:
                c, a = parts[0], parts[1]
                # very crude heuristic
                c_capitalized = c.capitalize()
                if c_capitalized in routes and a.capitalize() not in [ac.capitalize() for ac in routes[c_capitalized]]:
                   # errors.append(f"Possible 404: Href {href} -> '{a}' not found in Controller '{c_capitalized}'. Referenced in {link['file']}")
                   pass
            
    for redirect in redirects:
        c = redirect['controller']
        a = redirect['action']
        if c not in routes:
            errors.append(f"404: Redirect Controller '{c}' not found. Referenced in {redirect['file']} (Action: {a})")
        elif a not in set([ac for ac in routes[c]]):
            errors.append(f"404: Redirect Action '{a}' not found in Controller '{c}'. Referenced in {redirect['file']}")

    for error in set(errors):
        print(error)

if __name__ == '__main__':
    analyze()
